### PR TITLE
liquidctl 1.3.2 (new formula)

### PR DESCRIPTION
**Pull request by CaseySJ**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
I would like to submit `liquidctl` for inclusion in the `homebrew-core` tap. As a moderator on the [TonyMac forum](https://www.tonymacx86.com/threads/success-gigabyte-designare-z390-thunderbolt-3-i7-9700k-amd-rx-580.267551/), I have run into a fair number of people who are interested in installing and running this tool to monitor and control their PC power supplies, liquid coolers, fans, and RGB LEDs. The PC enthusiast community is still vibrant, and they're looking for solutions such as `liquidctl` to provide -- for the first time -- the ability to perform these functions on their custom builds. 

I have found `liquidctl` to be exceptionally well written and well maintained. It has grown over the past year to provide support for an expanded number of devices, and is proving itself to be an increasingly valuable tool for the enthusiast community.

Please consider adding it to `homebrew-core` so an even greater number of people will use homebrew for the first time, and be able to install `liquidctl` with a single `brew install liquidctl` command. Thank you.

### CODE REVIEW

**Comment by SMillerDev** [comment]
All versions in homebrew-core need to be stable. Please don't add development builds.

**Comment by jonasmalacofilho** [comment]
I think this has already been fixed, unless you mean that the `head` URL is no longer allowed as well.

**Comment by SMillerDev** [comment]
The title says devel and the formula has that version so I assumed that was a development version. If I was wrong you can change the title and ignore that comment

**Comment by CaseySJ** [comment]
I've changed the title to simply `liquidctl 1.3.0 (new formula)`. Although this pull request started with 1.2.0, a newer 1.3.0 version is ready.

File: Formula/liquidctl.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  resource "pyusb" do
+    url "https://files.pythonhosted.org/packages/source/p/pyusb/pyusb-1.0.2.tar.gz"
+    sha256 "4e9b72cc4a4205ca64fbf1f3fff39a335512166c151ad103e55c8223ac147362"
+  end
+
+  def install
+    # customize liquidctl --version
+    ENV["DIST_NAME"] = "homebrew"
+    ENV["DIST_PACKAGE"] = "#{tap.nil? ? path : full_name} #{version}"
```
Comment (SMillerDev) [inline Formula/liquidctl.rb]: Why is this needed?

File: Formula/liquidctl.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    man_page = buildpath/"liquidctl.8"
+    # setting the is_macos register to 1 adjusts the man page for macOS
+    inreplace man_page, ".nr is_macos 0", ".nr is_macos 1"
+    man.mkpath
+    man8.install man_page
+  end
+
+  test do
+    shell_output "#{bin}/liquidctl list --verbose --debug"
```
Comment (SMillerDev) [inline Formula/liquidctl.rb]: We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

File: Formula/liquidctl.rb (hunk 1 above)
Comment (jonasmalacofilho) [reply Formula/liquidctl.rb]: Upstream developer here...

This customizes the output of `liquidctl --version` adding some downstream package information so I know that I can (sort of) trust the version information and track down any build/dependency/platform issues.

It's by no means required, but it's nice to have (for me).

File: Formula/liquidctl.rb (hunk 2 above)
Comment (jonasmalacofilho) [reply Formula/liquidctl.rb]: This actually tests a _lot_ of stuff, as it runs the entire probe pipeline.  If a dependency is missing or if there's something majorly broken this will exit with non-zero status.

It's the best that can be done without assuming the availability of any specific hardware device.

---

Unit tests already run in CI for every upstream commit.

File: Formula/liquidctl.rb (hunk 1 above)
Comment (jonasmalacofilho) [reply Formula/liquidctl.rb]: And the conditional is needed so the formula can still be tested locally/directly from the file.

```rb
# The {Tap} instance associated with this {Formula}.
# If it's <code>nil</code>, then this formula is loaded from path or URL.
# @private
attr_reader :tap
```

https://github.com/Homebrew/brew/blob/d65ad9fe1a96fd895d24e097a3d69b9d280ee902/Library/Homebrew/formula.rb#L96-L99

Comment (SMillerDev) [reply Formula/liquidctl.rb]: That seems like a lot of code that only happens when you write the formula the first time.

Comment (jonasmalacofilho) [reply Formula/liquidctl.rb]: I'm not sure, but wouldn't the formula be installed from the local file when testing an updated version?

Anyway, this is your area of expertise and

```
ENV["DIST_PACKAGE"] = "#{full_name} #{version}"
```

would be fine be me.

File: Formula/liquidctl.rb (hunk 2 above)
Comment (CaseySJ) [reply Formula/liquidctl.rb]: To add a little more detail to this thread...

`liquidctl` monitors and controls fan speeds, voltages, and RGB lighting effects in a growing number of PC components such as all-in-one liquid coolers, fan controllers, smart power supplies, and lighting controllers. 

If the host computer does not contain a compatible device, then of course `liquidctl` is unable to function.

Therefore the test block in the Formula checks that the entire infrastructure is installed and ready should the user add a compatible device to their system.

File: Formula/liquidctl.rb (hunk 1 above)
Comment (jonasmalacofilho) [reply Formula/liquidctl.rb]: @SMillerDev, you never answered: wouldn't a user trying to update _and test_ the formula need to install it directly from the local file?

@CaseySJ feel free to remove DIST_NAME/DIST_PACKAGE.

Comment (SMillerDev) [reply Formula/liquidctl.rb]: No, actually you don't even need to for writing it the first time. You can just edit the file in `$(brew --prefix homebrew/core)/Formulae/liquidctl.rb` and test it using `brew test liquidctl`.

Comment (jonasmalacofilho) [reply Formula/liquidctl.rb]: @SMillerDev oh, good to know, thanks!

So would

```
    # customize liquidctl --version
    ENV["DIST_NAME"] = "homebrew"
    ENV["DIST_PACKAGE"] = "liquidctl #{version}"
```

be acceptable or should Casey remove everything?

Comment (SMillerDev) [reply Formula/liquidctl.rb]: This makes sense to keep. Although the package might be a little redundant? The liquidctl package will always be liquidctl?

Comment (jonasmalacofilho) [reply Formula/liquidctl.rb]: The idea is that `DIST_PACKAGE` should match the user visible name of the actual package.  So, _in homebrew-core,_ it should always be `liquidctl #{version}`.

But, for comparison, in the custom tap I use `#{full_name} #{version}`, and on Linux distros this can vary wildly and include other important information like build version.

---

On liquidctl side...

By default `liquidctl --version` will only output the base/hardcoded version (e.g. `liquidctl v1.3.1`).

But if both `DIST_NAME` and `DIST_PACKAGE` are passed, the output will then include this additional information.  An example would be: `liquidctl v1.3.1 (Arch Linux; python-liquidctl-git 1.3.1.r1.8e67646-1)`.

If manually installed from sources, with git, it will include the commit hash.  And if installed in pip --editable mode, it will instead just make a note of that.

Comment (jonasmalacofilho) [reply Formula/liquidctl.rb]: @CaseySJ I think Sean meant that the cleaner `ENV["DIST_PACKAGE"] = "liquidctl #{version}"` made sense to keep.

Comment (CaseySJ) [reply Formula/liquidctl.rb]: Understood. Will update the formula this evening and push the branch again.

**Comment by jonasmalacofilho** [comment]
@SMillerDev can you please review this again, after the latest changes?

**Comment by SMillerDev** [comment]
Could you squash all your commits into one `liquidctl 1.3.2 (new formula)`?

**Comment by CaseySJ** [comment]
> Could you squash all your commits into one `liquidctl 1.3.2 (new formula)`?

Certainly -- commits have been squashed per the Brew Docs:
```
git checkout <YOUR_BRANCH_NAME>
git rebase --interactive origin/master
git push --force
```

**Comment by jonasmalacofilho** [comment]
@CaseySJ during the `--interactive` rebase you can _specify_ that you want to squash some commits (all but the first one) and reword the resulting commit message.

**Comment by jonasmalacofilho** [comment]
@CaseySJ

One last thing (I think): the commit message is supposed to read `liquidctl 1.3.2 (new formula)`, subject only (see Sean's [comment](https://github.com/Homebrew/homebrew-core/pull/46442#issuecomment-565019062)).

You can fix this with another interactive rebase or, since there's only one commit, with

```
git commit --amend
```

**Comment by CaseySJ** [comment]
> @CaseySJ
> 
> One last thing (I think): the commit message is supposed to read `liquidctl 1.3.2 (new formula)`, subject only (see Sean's [comment](https://github.com/Homebrew/homebrew-core/pull/46442#issuecomment-565019062)).
> 
> You can fix this with another interactive rebase or, since there's only one commit, with
> 
> ```
> git commit --amend
> ```
> 
> After that, just force push again.
Changes made and pushed.

**Comment by jonasmalacofilho** [comment]
LGTM, thanks again @CaseySJ.

**Comment by jonasmalacofilho** [comment]
Ping @SMillerDev

**Comment by SMillerDev** [comment]
Can you also make it say `(new formula)` in the commit message?
